lib/render/ParticleRenderer.cpp

```diff
@@ -0,0 +1,163 @@
+#include <string>
+#include <iterator>
+
+#include <vapor/glutil.h>    // Must be included first!!!
+
+#include <vapor/Renderer.h>
+#include <vapor/ParticleRenderer.h>
+#include <vapor/ParticleParams.h>
+#include <vapor/regionparams.h>
+#include <vapor/ViewpointParams.h>
+#include <vapor/DataStatus.h>
+#include <vapor/DataMgrUtils.h>
+#include <vapor/errorcodes.h>
+#include <vapor/ControlExecutive.h>
+#include "vapor/ShaderManager.h"
+#include "vapor/debug.h"
+#include <glm/glm.hpp>
+#include <glm/gtc/type_ptr.hpp>
+#include "vapor/GLManager.h"
+#include <vapor/LegacyGL.h>
+
+using namespace VAPoR;
+
+#pragma pack(push, 4)
+// struct ParticleRenderer::VertexData {
+//    float x, y, z;
+//    float v;
+//};
+#pragma pack(pop)
+
+static RendererRegistrar<ParticleRenderer> registrar(ParticleRenderer::GetClassType(), ParticleParams::GetClassType());
+
+ParticleRenderer::ParticleRenderer(const ParamsMgr *pm, string winName, string dataSetName, string instName, DataMgr *dataMgr)
+: Renderer(pm, winName, dataSetName, ParticleParams::GetClassType(), ParticleRenderer::GetClassType(), instName, dataMgr)
+{
+}
+
+ParticleRenderer::~ParticleRenderer() {}
+
+int ParticleRenderer::_paintGL(bool)
+{
+    int rc = 0;
+    glDepthMask(true);
+    glEnable(GL_DEPTH_TEST);
+
+    auto p = GetActiveParams();
+
+    MapperFunction *mf = p->GetMapperFunc(p->GetVariableName());
+    float           LUT[256 * 4];
+    mf->makeLut(LUT);
+    float mapMin = mf->getMinMapValue();
+    float mapMax = mf->getMaxMapValue();
+    float mapDif = mapMax - mapMin;
+
+    vector<double> minExt, maxExt;
+    p->GetBox()->GetExtents(minExt, maxExt);
+#define PD3(v) printf("%s = %f, %f, %f\n", #v, v[0], v[1], v[2])
+    string varName = p->GetVariableName();
+    Grid * grid = _dataMgr->GetVariable(p->GetCurrentTimestep(), varName, p->GetRefinementLevel(), p->GetCompressionLevel(), minExt, maxExt);
+    if (!grid) return -1;
+
+    vector<long> values;
+
+    size_t        stride = max(1L, p->GetValueLong(ParticleParams::StrideTag, 1));
+    bool          showDir = p->GetValueLong(ParticleParams::ShowDirectionTag, 0);
+    float         dirScale = p->GetValueDouble(ParticleParams::DirectionScaleTag, 1);
+    vector<Grid *> vecGrids;
+    if (showDir) {
+        vector<string> vecNames = p->GetFieldVariableNames();
+        vector<string> mainVarCoords;
+        _dataMgr->GetVarCoordVars(varName, true, mainVarCoords);
+
+        for (auto var : vecNames) {
+            vector<string> varCoords;
+            _dataMgr->GetVarCoordVars(var, true, varCoords);
+
+            if (mainVarCoords != varCoords) {
+                if (grid) delete grid;
+                for (auto g : vecGrids) delete g;
+                SetErrMsg("Variable \"%s\" on different grid from main variable", var.c_str());
+                return -1;
+            }
+
+            Grid *ng = _dataMgr->GetVariable(p->GetCurrentTimestep(), var, p->GetRefinementLevel(), p->GetCompressionLevel(), minExt, maxExt);
```

Comment:
Better to use DataMgrUtils::GetGrids which will fetch multiple variables at once and temporarily "lock" them so that fetching one variable doesn't purge another one from the cache. Alternatively, you can use DataMgr::GetVariable() with lock=true, but then you'll need to explicitly unlock them.

Reply:
Done